### PR TITLE
fix: allow "localhost" in `cy.intercept`'s RouteMatcher

### LINKED ISSUE #17015: v7.4.0 No longer allows localhost as a RouteMatcher hostname
### Current behavior

`cy.intercept({ hostname: 'localhost' }, (req) => {
...
})`

Results in

```
An invalid RouteMatcher was supplied to cy.intercept(). hostname must be a valid host name or domain name.

You passed: {
  "hostname": "localhost"
}
```

Screen Shot 2021-06-18 at 1 55 53 PM

### Desired behavior

As the docs suggest (https://docs.cypress.io/api/commands/intercept#Usage), I should be able to use 'localhost' as the `hostname`.

>
```
// spying, dynamic stubbing, request modification, etc.
cy.intercept('/users*', { hostname: 'localhost' }, (req) => {
  /* do something with request and/or response */
})
```

### Test code to reproduce

`cy.intercept({ hostname: 'localhost' }, (req) => {
...
})`

### Versions

Introduced in 7.4.0 in this PR https://github.com/cypress-io/cypress/pull/16577

Worked in 7.3.0

### PR DESCRIPTION
**Pull request by sainthkh**
- Closes #17015

### User facing changelog

Allow "localhost" in `cy.intercept`'s RouteMatcher.

### Additional details
- Why was this change necessary? -> `localhost` is a valid hostname. But it became an error.
- What is affected by this change? -> N/A
- Any implementation details to explain? -> N/A

### How has the user experience changed?

N/A

### PR Tasks
- [x] Have tests been added/updated?

### CODE REVIEW

**Comment by sainthkh** [comment]
flaky failures.

**Comment by jennifer-shehane** [comment]
@sainthkh Zach is out of town for the week, so this review will be delayed til he gets back. I think he mentioned that this should be expanded a bit to include more than just "localhost", but I'm not sure exactly what he meant by that.

File: packages/driver/src/cy/net-stubbing/add-command.ts [hunk 1]
```diff
@@ -131,7 +131,7 @@ function validateRouteMatcherOptions (routeMatcher: RouteMatcherOptions): { isVa
     }
   }
 
-  if (_.isString(routeMatcher.hostname) && !isValidDomain(routeMatcher.hostname, { allowUnicode: true })) {
+  if (_.isString(routeMatcher.hostname) && !isValidDomain(routeMatcher.hostname, { allowUnicode: true }) && routeMatcher.hostname !== 'localhost') {
```
Comment (flotwig) [inline packages/driver/src/cy/net-stubbing/add-command.ts:134]: This change is not comprehensive enough, I think any hostname should be allowed to work here. For example, `somemachine`, `local6`, `router`, etc. should all work.

You could try to patch `isValidDomain`: https://github.com/miguelmota/is-valid-domain/issues/23

Alternatively, you could kludge it here, by checking if `hostname` contains a `.`, and if not, appending `.sometld` before passing it to `isValidDomain`...

**Review by flotwig — CHANGES_REQUESTED** [review]

File: packages/driver/src/cy/net-stubbing/add-command.ts (hunk 1 above)
Comment (flotwig) [reply packages/driver/src/cy/net-stubbing/add-command.ts:134]: or https://github.com/miguelmota/is-valid-hostname, of course

**Review by flotwig — APPROVED** [review]

**Pull request merged**